server/src/main/java/org/elasticsearch/cluster/health/ClusterStateHealth.java

```diff
@@ -144,6 +151,44 @@ public ClusterStateHealth(
         this.indices = indices;
     }
 
+    public static ClusterHealthStatus getHealthStatus(
+        final RoutingTable routingTable,
+        final RoutingNodes routingNodes,
+        final Logger logger
+    ) {
+        ClusterHealthStatus computeStatus;
+        if (routingNodes.unassigned().size() == 0 && routingNodes.hasInactiveShards() == false) {
+            // no unassigned, inactive shards
+            return ClusterHealthStatus.GREEN;
+        }
+        if (routingNodes.hasUnassignedPrimaries() == false && routingNodes.hasInactivePrimaries() == false) {
+            // no unassigned, inactive primaries, but has unassigned or inactive replicas
+            return ClusterHealthStatus.YELLOW;
+        }
+
+        // cluster at least YELLOW
+        computeStatus = ClusterHealthStatus.YELLOW;
+        for (IndexRoutingTable indexRoutingTable : routingTable) {
```

Comment:
Practically, we might end up in this loop quite a bit won't we?  I wonder if it's worth it computing `RoutingNodes` for one of the states (the other already has them) here if that's the case.

Maybe we could do a smaller change here and do this PR without the short-cuts through `RoutingNodes` and take the win for not having to create the `ClusterIndexHealth` etc. instances and map of them here for now and do without the routing nodes for a straight-forward speedup in all cases?
So concretely, always do this loop but start with green health and do the same loop logically that we do in `org.elasticsearch.cluster.health.ClusterStateHealth#ClusterStateHealth(org.elasticsearch.cluster.ClusterState, java.lang.String[])`.

To me it it seems like if we wanted to use `RoutingNodes` here (again, I removed its used a short while ago) we'd just want to compute the index health when building the routing nodes directly and make this method free.
WDYT?

Reply:
I thought `RoutingNodes` has been constructed before. I am not sure I got all the parts of you suggestion, I have updated the PR to show the code, please help to check again, thank you.

Reply:
Thanks! Sorry if I was unclear here, turned out exactly as I had hoped though :)
We don't always have the `RoutingNodes` here for the new cluster state. We probably will have to build them later anyway, but it's a lot easier for me to reason about the performance of this if we use the routing table :)